Utilities/WDLs/Dipcall.wdl

```diff
@@ -0,0 +1,270 @@
+            cp ~{assemblyFastaMat} .
+            gunzip $MAT_FILENAME
+            MAT_FILENAME="${MAT_FILENAME%.gz}"
+        else
+            ln -s ~{assemblyFastaMat}
+        fi
+
+        # prep reference
+        REF_FILENAME=$(basename -- "~{referenceFasta}")
+        if [[ $REF_FILENAME =~ \.gz$ ]]; then
+            cp ~{referenceFasta} .
+            gunzip $REF_FILENAME
+            REF_FILENAME="${REF_FILENAME%.gz}"
+        else
+            ln -s ~{referenceFasta}
+        fi
+        samtools faidx $REF_FILENAME
+
+        # initialize script
+        cmd=( /opt/dipcall/dipcall.kit/run-dipcall )
+
+        # male samples need PAR region excluded
+        if [[ ~{isMaleSample} == true ]]; then
+            if [[ ~{referenceIsHS38} ]]; then
+                cmd+=( -x /opt/dipcall/dipcall.kit/hs38.PAR.bed )
+            elif [ -f "~{custom_PAR_bed}" ]; then
+                cmd+=( -x "~{custom_PAR_bed}" )
+            else
+                echo "WARNING: Assuming reference is hg37 because input is not hg38 and no custom PAR provided."
+                echo "Using built-in PAR regions for hg37 to exclude from calling..."
+                cmd+=( -x /opt/dipcall/dipcall.kit/hs37d5.PAR.bed )
+            fi
+
+        fi
+
+        # finalize script
+        cmd+=( $PREFIX.dipcall/$PREFIX )
+        cmd+=( $REF_FILENAME )
+        cmd+=( $PAT_FILENAME )
+        cmd+=( $MAT_FILENAME )
+
+        # generate makefile
+        "${cmd[@]}" >$PREFIX.mak
+
+        # run dipcall
+        make -j 2 -f $PREFIX.mak
+
+        # finalize
+        rm $PREFIX.dipcall/*sam.gz
+        tar czvf $PREFIX.dipcall.tar.gz $PREFIX.dipcall/
+        cp $PREFIX.dipcall/$PREFIX.dip.bed $PREFIX.dipcall.bed
+        cp $PREFIX.dipcall/$PREFIX.dip.vcf.gz $PREFIX.dipcall.vcf.gz
+
+        # cleanup
+        rm $REF_FILENAME
+        rm $MAT_FILENAME
+        rm $PAT_FILENAME
+
+	>>>
+	output {
+		File outputTarball = glob("*.dipcall.tar.gz")[0]
+		File outputVCF = glob("*.dipcall.vcf.gz")[0]
+		File outputBED = glob("*.dipcall.bed")[0]
+	}
+    runtime {
+        memory: memSizeGB + " GB"
+        cpu: threadCount
+        disks: "local-disk " + diskSizeGB + " SSD"
+        docker: dockerImage
+        preemptible: 1
+    }
+}
+
+# Task assumes contig names in PAR bed match the allosome names in Dipcall VCF, but otherwise name agnostic
+# Also assumes that PAR are on the edges of X/Y with two connected components and nowhere else, as in humans
+task CleanHaploidCalls {
+    input {
+        File dipcall_vcf
+        File PAR_bed
+
+        # Runtime arguments
+        Int disk_size = 2*ceil(size(dipcall_vcf, "GB")) + 50
+        Int cpu = 4
+        Int memory = 8
+    }
+
+    command <<<
+        set -xueo pipefail
+
+        python << CODE
+
+        import pysam
+        from cyvcf2 import VCF, Writer
+
+
+        input_file = "~{dipcall_vcf}"
+        bed_file = "~{PAR_bed}"
+
+        pysam.tabix_index(input_file, preset='vcf', force=True)
```

Comment:
In this case the input file won't have an index (the previous task doesn't create it), but in fact I don't think the index is needed for this task since it iterates over the entire VCF. This is probably a holdover from some testing out on small regions long ago. I removed it from the script.

Reply:
Marie Kando liked this.